File: 003-algorithms/002-sorting/exercise/006-heap-sort/solution.py
"""
 * Approach: Heap Sort
 * Time complexity: O(n log n)
 * Space complexity: O(1)
"""


def heap_sort(A):
    heap_size = len(A)
    build_heap(A, heap_size)
    for i in range(heap_size-1, 0, -1):
        A[0], A[i] = A[i], A[0]
        heap_size -= 1
        max_heapify(A, heap_size, 0)
    return A


def build_heap(A, heap_size):
    for i in range((heap_size//2), -1, -1):
        max_heapify(A, heap_size, i)


def max_heapify(A, heap_size, i):
    left = 2 * i + 1
    right = 2 * i + 2
    largest = i
    if left < heap_size and A[left] > A[largest]:
        largest = left
    if right < heap_size and A[right] > A[largest]:
        largest = right
    if largest != i:
        A[i], A[largest] = A[largest], A[i]
        max_heapify(A, heap_size, largest)


print(heap_sort([99, 44, 6, 2, 1, 5, 63, 87, 283, 4, 0]))

File: 003-algorithms/002-sorting/exercise/006-heap-sort/solution.js
/**
 * Approach: Heap Sort
 * Time complexity: O(n log n)
 * Space complexity: O(1)
 */

function heapSort(array) {
  let heapSize = array.length;
  buildHeap(array, heapSize);

  for (let i of range(heapSize - 1, 0, -1)) {
    [array[0], array[i]] = [array[i], array[0]];
    heapSize -= 1;
    maxHeapify(array, heapSize, 0);
  }

  return array;
}

function buildHeap(array, heapSize) {
  for (let i of range(Math.floor(heapSize / 2), -1, -1)) {
    maxHeapify(array, heapSize, i);
  }
}

function* range(start, stop, step = 1) {
  if (stop === undefined) {
    stop = start;
    start = 0;
  }

  for (let i = start; step > 0 ? i < stop : i > stop; i += step) {
    yield i;
  }
}

function maxHeapify(array, heapSize, i) {
  let left = 2 * i + 1;
  let right = 2 * i + 2;
  let largest = i;

  if (left < heapSize && array[left] > array[largest]) {
    largest = left;
  }

  if (right < heapSize && array[right] > array[largest]) {
    largest = right;
  }

  if (largest !== i) {
    [array[i], array[largest]] = [array[largest], array[i]];
    maxHeapify(array, heapSize, largest);
  }
}

console.log(heapSort([99, 44, 6, 2, 1, 5, 63, 87, 283, 4, 0]));
